hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepRunner.java

```diff
@@ -99,10 +99,24 @@ public static boolean isStepRunnerCurrentlyExecutingOnPartitionThread() {
         return CURRENTLY_EXECUTING_ON_PARTITION_THREAD.get();
     }
 
-    private void addOpToOffloadedOps(MapOperation op) {
-        op.setOperationResponseHandler(new OffloadedStepResponseHandler(op.getOperationResponseHandler()));
-        offloadedOperations.add(op);
-        op.getRecordStore().incMapStoreOffloadedOperationsCount();
+    private boolean addOpToOffloadedOps(MapOperation op) {
+        if (offloadedOperations.add(op)) {
+            op.getRecordStore().incMapStoreOffloadedOperationsCount();
```

Comment:
How about getting rid of this? We can return `offloadedOperations.size()` in `RecordStore.getMapStoreOffloadedOperationsCount()`?

Reply:
Reason is thread safety, offloadedOperations is not a thread-safe list but counter is thread-safe. Metrics thread is reading this counter but offloadedOperations is designed to be accessed only single partition thread.